pkg/controller/messaginginfra/messaginginfra_controller.go

```
@@ -0,0 +1,216 @@
+	}
+
+	// Reconcile CA
+	err = ic.ProcessSimple(func(infra *v1beta2.MessagingInfra) error {
+		return r.certController.ReconcileCa(ctx, reqLogger, infra)
+	})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Reconcile Routers
+	err = ic.ProcessSimple(func(infra *v1beta2.MessagingInfra) error {
+		return r.routerController.ReconcileRouters(ctx, reqLogger, infra)
+	})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Reconcile Brokers
+	err = ic.ProcessSimple(func(infra *v1beta2.MessagingInfra) error {
+		return r.brokerController.ReconcileBrokers(ctx, reqLogger, infra)
+	})
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Update main condition
+	err = ic.ProcessSimple(func(infra *v1beta2.MessagingInfra) error {
+		routersCreated := infra.Status.GetMessagingInfraCondition(v1beta2.MessagingInfraRoutersCreated)
+		brokersCreated := infra.Status.GetMessagingInfraCondition(v1beta2.MessagingInfraBrokersCreated)
+		caCreated := infra.Status.GetMessagingInfraCondition(v1beta2.MessagingInfraCaCreated)
+
+		// TODO: Update ready based on actual router and broker state
+		ready := infra.Status.GetMessagingInfraCondition(v1beta2.MessagingInfraReady)
+		if routersCreated.Status == corev1.ConditionTrue &&
+			brokersCreated.Status == corev1.ConditionTrue &&
+			caCreated.Status == corev1.ConditionTrue {
+
+			ready.SetStatus(corev1.ConditionTrue, "", "")
+			infra.Status.Phase = v1beta2.MessagingInfraActive
+
+		} else {
+			ready.SetStatus(corev1.ConditionFalse, "", "Infrastructure is not created")
+		}
+		return nil
+	})
+
+	return reconcile.Result{}, nil
+}
+
+/*
+ * Automatically handle status update of the infra after running processor.
+ */
+type infraContext struct {
+	ctx    context.Context
+	client client.Client
+	infra  *v1beta2.MessagingInfra
+	status *v1beta2.MessagingInfraStatus
+}
+
+func (i *infraContext) Process(processor func(infra *v1beta2.MessagingInfra) (reconcile.Result, error)) (reconcile.Result, error) {
```

Calling this for every step would cause a huge amount of changes. I am also not quite sure if you can call update twice on the same object. I think the optimistic lock would fail, at least that is what happens when you update the object itself.

---

Hmm, good point. The intention was that status gets updated so that a user can see if an error has occured and/or what changes have actually passed, but I see the problem you're describing.

So I propose that I change this to the following:

If an error is returned, update status with latest changes
If error is not returned, accumulate status changes until all reconcile loops have changed.

I think this is similar to the work you recently did.